pkg/workloads/cortex/serve/serve.py

```diff
@@ -206,27 +210,45 @@ def predict(request: Request):
                 api.monitoring.model_type == "classification"
                 and predicted_value not in local_cache["class_set"]
             ):
-                tasks = BackgroundTasks()
                 tasks.add_task(api.upload_class, class_name=predicted_value)
```

Comment:
What order do these tasks run? If the post predict fails, does this task get affected? Should we run post_predict after recording this metric?

Reply:
Good observation.

The tasks are executed sequentially, and if one of them fails, all the rest in the list are skipped too.

In this case, just like you have suggested, I think it makes more sense to place `post_predict` after `upload_class` because I wouldn't want to add extra logic to the critical path of the request to handle exceptions (i.e. wrap `post_predict` within a try/except block in a placeholder function). This extra logic can reduce the maximum potential throughput of the application.

I also think adding it afterward makes more sense because the response also gets validated before calling the `post_predict` method - the thinking is that by the time `post_predict` is called, you should already assume the `response` param of `post_predict` method is a valid response of `predict` method.